RFC-0004-functions-stats.md

```diff
@@ -0,0 +1,561 @@
+    USE_MAX_ARGUMENT,
+    /** Sum the stats value of all arguments to derive the new stats value */
+    SUM_ARGUMENTS,
+    /** Propagate the source stats as-is */
+    USE_SOURCE_STATS,
+    /** Use the value of output row count. */
+    ROW_COUNT,
+    /** Use the value of row_count * (1 - null_fraction). */
+    NON_NULL_ROW_COUNT,
+    /** use the value of TYPE_WIDTH in varchar(TYPE_WIDTH) */
+    USE_TYPE_WIDTH_VARCHAR,
+    /** Take max of type width of arguments with varchar type. */
+    MAX_TYPE_WIDTH_VARCHAR,
+    /** Stats are unknown and thus no action is performed. */
+    UNKNOWN
+}
+```
+
+`FunctionMetadata` is expanded to include stats, which is available for processing at
+[ScalarStatsCalculator](https://github.com/prestodb/presto/blob/0.289-edge7/presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java#L124)
+
+```java
+public class FunctionMetadata
+{
+    // existing fields...
+    private final Map<Signature, ScalarStatsHeader> statsHeader;
+    // ...
+}
+```
+
+Where `ScalarStatsHeader` looks like:
+
+```java
+public class ScalarStatsHeader
+{
+    private Map<Integer, ScalarPropagateSourceStats> argumentStatsResolver;
+    private double distinctValuesCount;
+    private double nullFraction;
+    private double avgRowSize;
+    private double min;
+    private double max;
+    //...
+    /*
+     * Get stats annotation for each of the scalar function argument, where key is the index of the position
+     * of functions' argument and value is the ScalarPropagateSourceStats annotation.
+     */
+    public Map<Integer, ScalarPropagateSourceStats> getArgumentStats()
+    {
+        return argumentStatsResolver;
+    }
+}
+```
+
+The field `argumentStatsResolver` is processed by parsing annotation `ScalarPropagateSourceStats` on functions. In the case of C++ implementation discussed in
+[phase 2](#phase-2-c-implementation) this is provided by `transformSourceStats` in `FunctionMetadata.h` and later embedded in json by `JsonBasedUdfFunctionMetadata`. 
+
+Examples of using above annotations for various function in `StringFunctions.java`
+
+1. upper(Slice)
+```java
+    @Description("converts the string to upper case")
+    @ScalarFunction
+    @LiteralParameters("x")
+    @SqlType("varchar(x)")
+    public static Slice upper(
+            @ScalarPropagateSourceStats @SqlType("varchar(x)") Slice slice)
+    {
+        return toUpperCase(slice);
+    }
+```
+Annotation: `ScalarPropagateSourceStats` is by default `propagateAllStats = true` i.e. the stats estimate for `upper(col1)` function will be same as the source column(i.e. `col1`)
+stats. 
+
+2. starts_with(Slice, Slice): an example of constant stats.
+```java
+    @SqlNullable
+    @Description("Returns whether the first string starts with the second")
+    @ScalarFunction("starts_with")
+    @LiteralParameters({"x", "y"})
+    @SqlType(StandardTypes.BOOLEAN)
+    @ScalarFunctionConstantStats(distinctValuesCount = 2, minValue = 0, maxValue = 1)
```

Comment:
"minValue = 0, maxValue = 1"
Not sure about this part for boolean value

Reply:
Hi @feilong-liu , thanks for taking a look. 
Yes, analyze on table with boolean column created under HIVE catalog shows NULL for min and max. So will fix this. 

P.S. You can checkout annotation for other functions on our POC PR: https://github.com/prestodb/presto/pull/22974